### PR TITLE
feat: create new amplify-prompts package to handle all terminal interactions

### PR DESCRIPTION
**Pull request by edwardfoyle**
#### Description of changes

Adds a new `amplify-prompts` package that can be used in place of the print object that hangs off of the amplify context object and can also replace our inconsistent usage of inquirer and enquirer across the codebase to provide a more consistent input experience moving forward.

`amplify-prompts` aims to provide an interface that makes common use cases such as simply asking for a string input or picking an item from a list dead simple. It also supports more complex use cases although it intentionally exposes a limited set of functionality of the underlying libraries in order to facilitate a more uniform experience across CLI prompts. All inputs and outputs are fully typed.

To get a sense for the interface, check out `demo.ts` (or pull the PR and run `yarn demo`)

This PR does NOT change any existing functionality (ie `amplify-prompts` is not used anywhere in the CLI yet). However, it should be used moving forward as we add / refactor code that uses prompts. It is not an exact drop-in replacement for inquirer / enquirer but it should require minimal modification of calling code to make a replacement.

The existing methods that this package replaces have been marked @deprecated to point developers to the new methods

#### Issue #, if available

Been working on this on the side for a few weeks now. Every time I've seen a PR that plumbs "context" through 5 function calls just to print a line to the console it makes me cry a little so I made a library to fix it. And hopefully this library also makes it easier / safer (in terms of types) to collect customer input in general.

#### Description of how you validated changes
Manual testing and near 100% unit test coverage

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#tests)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-prompts/package.json [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    "prompts"
+  ],
+  "author": "Amazon Web Services",
+  "license": "Apache-2.0",
+  "bugs": {
+    "url": "https://github.com/aws-amplify/amplify-cli/issues"
+  },
+  "homepage": "https://github.com/aws-amplify/amplify-cli#readme",
+  "dependencies": {
+    "chalk": "^4.1.1",
```
Comment (jhockett) [inline packages/amplify-prompts/package.json:29]: is it possible to update chalk in the rest of the packages to match 4.1.1? Or is it something we need to do over time once we refactor enough code to use this package over context.print.x?

File: packages/amplify-prompts/src/__tests__/prompter.test.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+  it('returns prompt response if no transformer present', async () => {
+    const result = 'this is the result';
+    prompt_mock.mockResolvedValueOnce({ result });
+    expect(await prompter.input('test message')).toEqual(result);
+  });
+
+  it('returns transformed response if transformer present', async () => {
+    const promptResponse = 'this is the result';
+    const transformedValue = 'transformmed valie';
```
Comment (jhockett) [inline packages/amplify-prompts/src/__tests__/prompter.test.ts:67]: typos? transformmed => transformed, valie => value

```suggestion
    const transformedValue = 'transformed value';
```

File: packages/amplify-prompts/src/prompter.ts [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+
+type ValidatorCast = (input: string) => string | true | Promise<string> | Promise<true>;
+
+type TransformOption<T> = {
+  transform: (value: string) => T | Promise<T>;
+};
+
+type MaybeTransformOption<T> = T extends string ? Partial<TransformOption<T>> : TransformOption<T>;
+
+type MultiselectOption<M extends PickType> = M extends 'many'
```
Comment (jhockett) [inline packages/amplify-prompts/src/prompter.ts:188]: Should it be `Multiselect` or `MultiSelect`? (Here and elsewhere)

File: packages/amplify-prompts/package.json (hunk 1 above)
Comment (edwardfoyle) [reply packages/amplify-prompts/package.json:29]: resolving this as it looks like https://github.com/aws-amplify/amplify-cli/pull/7823 will take care of it

File: packages/amplify-prompts/src/prompter.ts (hunk 3 above)
Comment (edwardfoyle) [reply packages/amplify-prompts/src/prompter.ts:188]: looking through enquirer's types / docs, they write it as MultiSelect, so I'll update to that

**Review by jhockett — APPROVED** [review]
LGTM, this is a big improvement!

File: packages/amplify-prompts/src/prompter.ts (hunk 3 above)
Comment (edwardfoyle) [reply packages/amplify-prompts/src/prompter.ts]: I ended up renaming it to `ReturnSize` and reusing it it the pick and input prompts

File: packages/amplify-prompts/src/__tests__/prompter.test.ts [hunk 4]
```diff
@@ -0,0 +1,104 @@
+import { prompter } from '../prompter';
+import { prompt } from 'enquirer';
+import * as flags from '../flags';
+
+jest.mock('../flags');
+
+type Writeable<T> = { -readonly [P in keyof T]: T[P] };
```
[the remark this answers is not in the document]
Comment (edwardfoyle) [reply packages/amplify-prompts/src/__tests__/prompter.test.ts:7]: It removes the readonly modifier from the key (in this case all of the keys)

File: packages/amplify-prompts/src/printer.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+  };
+
+  private writeSilenceableLine = (line?: string): void => {
+    if (!isSilent) {
+      this.writeLine(line);
+    }
+  };
+
+  private writeLine = (line: string = ''): void => {
+    this.outputStream.write(`${line}\n`);
```
[the remark this answers is not in the document]
Comment (edwardfoyle) [reply packages/amplify-prompts/src/printer.ts:43]: I think most modern terminals on windows support just \n, but I can use os.EOL to be safe

**Review by jhockett — APPROVED** [review]

**Pull request merged**